### PR TITLE
feat: Service count can go up to 1024 in Docker backend

### LINKED ISSUE #912: Increase IP range of subnetworks
### Background & motivation

Right now the range of services is too narrow. This is due to some corner cutting on the code that assign IPs to enclaves.

### Desired behaviour

Increase the range, making a tighter calculation of IP bits

### How important is this to you?

Painful; the lack of this feature makes using Kurtosis frictionful.

### PR DESCRIPTION
**Pull request by victorcolombo**
## Is this change user facing?
NO

## References (if applicable):

Closes https://github.com/kurtosis-tech/kurtosis/issues/912

### CODE REVIEW

File: container-engine-lib/lib/backend_impls/docker/docker_network_allocator/docker_network_allocator.go [hunk 1]
```diff
@@ -142,14 +146,16 @@ func (provider *DockerNetworkAllocator) CreateNewNetwork(
...
 }
 
-// This algorithm only picks a network in the 172.16.1.0/24 - 17.16.0.255.0/24 list
+// This algorithm picks a network in the 172.16.0.0/16 range
 // https://github.com/hashicorp/serf/issues/385#issuecomment-208755148 - we try to follow RFC 6890
 // https://www.rfc-editor.org/rfc/rfc6890.html calls 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16 Private-Use (docker usually picks from 172.16.0.0/12)
-// Just with this range we can get 256 APICs; if we limit services per APIC to 256(/24)
-// For simplicity we limit it to 256 APICs and allow networks 172.16.1.0/24 - 17.16.255.0/24
+// We take IPs from the range 170.16.0.0/16, split equally into 2^6 networks, so the mask of a given network is:
+// 170.16.nnnnnn00.0/22 where nnnnnn are 6 bits to address the network, and the other 8 bits address the services
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/docker/docker_network_allocator/docker_network_allocator.go]: are there 8 or 10 bits for services?! just checking; above we write 2^10 ips per network and here we wright 8 bits

File: container-engine-lib/lib/backend_impls/docker/docker_network_allocator/docker_network_allocator_test.go [hunk 2]
```diff
@@ -25,31 +25,15 @@ func TestErrorOnNoFreeIps(t *testing.T) {
...
-		}
-		allPossibleNetworks = append(allPossibleNetworks, resultNetwork)
-	}
-	i := 0
-	for {
-		if i == thirdOctetHighestPossibleValue {
-			break
-		}
+	for i := 0; i < (1 << enclaveWidthBits); i++ {
 		freeIPAddress, err := findRandomFreeNetwork(takenNetworks)
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/docker/docker_network_allocator/docker_network_allocator_test.go:37]: can we assert that 1024 ip addresses are available for the 64 different networks that are possible?

**Review by h4ck3rk3y — APPROVED** [review]

File: container-engine-lib/lib/backend_impls/docker/docker_network_allocator/docker_network_allocator_test.go (hunk 2 above)
Comment (victorcolombo) [reply container-engine-lib/lib/backend_impls/docker/docker_network_allocator/docker_network_allocator_test.go:37]: Done!

File: container-engine-lib/lib/backend_impls/docker/docker_network_allocator/docker_network_allocator_test.go [hunk 3]
```diff
@@ -25,34 +26,29 @@ func TestErrorOnNoFreeIps(t *testing.T) {
...
+		require.NotContains(t, takenNetworks, freeIPAddress)
 		takenNetworks = append(takenNetworks, freeIPAddress)
-		require.Equal(t, allPossibleNetworks[i], freeIPAddress)
-		i += 1
+		require.EqualValues(t, 1<<networkWidthBits, countIPAddresses(freeIPAddress))
 	}
 }
 
+func countIPAddresses(ipNet *net.IPNet) int64 {
+	ones, bits := ipNet.Mask.Size()
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/docker/docker_network_allocator/docker_network_allocator_test.go]: perhaps totalBits and the difference can be numberOfFreeBits?

**Pull request merged**